dags/common/utils.py

```diff
@@ -321,37 +320,65 @@ def create_or_update_article(data):
-        if re.search(r"\b%s\b" % key, affiliation_value, flags=re.IGNORECASE):
-            return val
-    country = affiliation_value.split(",")[-1].strip()
-    for key, val in COUNTRIES_DEFAULT_MAPPING.items():
-        if re.search(r"\b%s\b" % key, country, flags=re.IGNORECASE):
-            return val
+def unwrap_coco_result(res):
+    if isinstance(res, (list, tuple)):
+        for item in res:
+            if isinstance(item, str) and item.lower() != "not found":
+                return item
+        return None
 
-    try:
-        country_code = cc.convert(country, to="iso2")
-        mapped_countries = []
-        if country_code != "not found":
-            mapped_countries = [
-                {
-                    "code": country_code,
-                    "name": cc.convert(country, to="name_short"),
-                }
-            ]
-
-        if len(mapped_countries) > 1 or len(mapped_countries) == 0:
-            raise FoundMoreThanOneMatchOrNone(affiliation_value)
-        return mapped_countries[0].get("name", "")
-    except (LookupError, FoundMoreThanOneMatchOrNone):
-        return find_country_match_from_mapping(affiliation_value)
-
-
-def find_country_match_from_mapping(affiliation_value):
-    for key in COUNTRIES_DEFAULT_MAPPING:
-        if re.search(r"\b%s\b" % key, affiliation_value, flags=re.IGNORECASE):
-            return COUNTRIES_DEFAULT_MAPPING[key]
+    if not res or str(res).lower() == "not found":
+        return None
+
+    return res
+
+
+def initialize_special_cases():
+    global SPECIAL_PATTERNS
+    sorted_keys = sorted(SPECIAL_CASES.keys(), key=len, reverse=True)
+    SPECIAL_PATTERNS = [
+        (re.compile(rf"\b{re.escape(key)}\b"), SPECIAL_CASES[key])
+        for key in sorted_keys
+    ]
+
+
+def parse_country_from_value(value):
+    if not SPECIAL_PATTERNS:
+        initialize_special_cases()
+
+    norm = value.lower().translate(CHAR_REPLACEMENTS)
+
+    for pattern, country in SPECIAL_PATTERNS:
+        if pattern.search(norm):
+            return country
+
+    raw_iso2 = cc.convert(names=norm, to="ISO2")
+    iso2 = unwrap_coco_result(raw_iso2)
+    if iso2:
+        raw_short = cc.convert(names=iso2, to="name_short")
+        short = unwrap_coco_result(raw_short)
+        if short:
+            return short
+
+    parts = norm.split()
+    plen = len(parts)
+    max_len = min(4, plen)
+    for i in range(plen):
+        for n in range(max_len, 0, -1):
+            start = plen - i - n
+            if start < 0:
+                continue
+            phrase = " ".join(parts[start : plen - i])
+            try:
+                matches = pycountry.countries.search_fuzzy(phrase)
+                if matches:
+                    return unwrap_coco_result(
+                        cc.convert(names=matches[0].alpha_2, to="name_short")
+                    )
+            except LookupError:
+                continue
+
+    return None
```

Comment:
and this one to a separate function, also I would add more descriptive names, for example `normilized_name_parts` `normilized_name_partslength`, `normilized_name_parts_length`